pkg/provisioners/gcppubsub/controller/channel/reconcile.go

```diff
@@ -178,20 +219,100 @@ func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel)
 		return false, err
 	}
 
-	topic, err := r.createTopic(ctx, c, gcpCreds, r.defaultGcpProject)
+	topic, err := r.createTopic(ctx, plannedPCS, gcpCreds)
 	if err != nil {
 		return false, err
 	}
 
-	err = r.createSubscriptions(ctx, c, gcpCreds, r.defaultGcpProject, topic)
+	err = r.syncSubscriptions(ctx, plannedPCS, gcpCreds, topic, subsToSync)
 	if err != nil {
 		return false, err
 	}
+	// Now that the subs have synced successfully, remove the old ones from the status.
+	plannedPCS.Subscriptions = subsToSync.subsToCreate
+	if err = pubsubutil.SaveRawStatus(ctx, c, plannedPCS); err != nil {
+		return false, err
+	}
 
 	c.Status.MarkProvisioned()
 	return false, nil
 }
 
+// planGcpResources creates the plan for every resource that needs to be created outside of the
+// cluster. Their names are all saved to the returned in the updated status.
+func (r *reconciler) planGcpResources(ctx context.Context, c *eventingv1alpha1.Channel, originalPCS *pubsubutil.GcpPubSubChannelStatus) (persistence, *pubsubutil.GcpPubSubChannelStatus, *syncSubs, error) {
+	persist := noNeedToPersist
+	subsToSync := &syncSubs{
+		subsToCreate: make([]pubsubutil.GcpPubSubSubscriptionStatus, 0),
+		subsToDelete: make([]pubsubutil.GcpPubSubSubscriptionStatus, 0),
+	}
+
+	topicName := originalPCS.Topic
+	if topicName == "" {
+		topicName = generateTopicName(c)
+	}
+	// Everything except the subscriptions.
+	newPCS := &pubsubutil.GcpPubSubChannelStatus{
+		// TODO Allow arguments to set the secret and project.
+		Secret:     r.defaultSecret,
+		SecretKey:  r.defaultSecretKey,
+		GCPProject: r.defaultGcpProject,
+		Topic:      topicName,
+	}
+	if !equality.Semantic.DeepEqual(originalPCS.Secret, newPCS.Secret) || originalPCS.SecretKey != newPCS.SecretKey || originalPCS.GCPProject != newPCS.GCPProject || originalPCS.Topic != newPCS.Topic {
+		persist = persistStatus
+	}
+
+	existingSubs := make(map[types.UID]pubsubutil.GcpPubSubSubscriptionStatus, len(originalPCS.Subscriptions))
+	for _, existingSub := range originalPCS.Subscriptions {
+		// I don't think this can ever happen, but let's just be sure.
+		if existingSub.Ref != nil && existingSub.Ref.UID != "" {
+			existingSubs[existingSub.Ref.UID] = existingSub
+		}
+	}
+	if c.Spec.Subscribable != nil {
+		for _, subscriber := range c.Spec.Subscribable.Subscribers {
+			if subscriber.Ref == nil || subscriber.Ref.UID == "" {
+				return noNeedToPersist, nil, nil, fmt.Errorf("empty reference UID: %v", subscriber)
```

Comment:
It doesn't make everything invalid, but it does simplify the logic in that either it all works or none of it does.

If you prefer, I can have it do all the work it can, leaving the bad Subscribers unfinished, and marking the `Channel` as not ready. Do you prefer that?

Reply:
I agree about emitting a corev1.Event and plan to do so in a subsequent PR.